pkg/control/pubcontrol/validate_pub.go

```diff
@@ -0,0 +1,160 @@
+	operation Operation, dryRun bool) (allowed bool, reason string, err error) {
+
+	// If the pod is not ready, it doesn't count towards healthy and we should not decrement
+	if !control.IsPodReady(pod) {
+		klog.V(3).Infof("pod(%s.%s) is not ready, then don't need check pub", pod.Namespace, pod.Name)
+		return true, "", nil
+	}
+
+	pub := control.GetPodUnavailableBudget()
+	// pod is in pub.Status.DisruptedPods or pub.Status.UnavailablePods, then don't need check it
+	if isPodRecordedInPub(pod.Name, pub) {
+		klog.V(5).Infof("pod(%s.%s) already is recorded in pub(%s.%s)", pod.Namespace, pod.Name, pub.Namespace, pub.Name)
+		return true, "", nil
+	}
+
+	pubClone := pub.DeepCopy()
+	refresh := false
+	err = retry.RetryOnConflict(ConflictRetry, func() error {
+		if refresh {
+			key := types.NamespacedName{
+				Name:      pubClone.Name,
+				Namespace: pubClone.Namespace,
+			}
+			if err := client.Get(context.TODO(), key, pubClone); err != nil {
+				klog.Errorf("Get PodUnavailableBudget(%s) failed: %s", key.String(), err.Error())
+				return err
+			}
+		}
+		// Try to verify-and-decrement
+		// If it was false already, or if it becomes false during the course of our retries,
+		err := checkAndDecrement(pod.Name, pubClone, operation)
+		if err != nil {
+			return err
+		}
+
+		// If this is a dry-run, we don't need to go any further than that.
+		if dryRun == true {
+			klog.V(5).Infof("pod(%s) operation for pub(%s.%s) is a dry run", pod.Name, pubClone.Namespace, pubClone.Name)
+			return nil
+		}
+		klog.V(3).Infof("pub(%s.%s) update status(disruptedPods:%d, unavailablePods:%d, expectedCount:%d, desiredAvailable:%d, currentAvailable:%d, unavailableAllowed:%d)",
+			pubClone.Namespace, pubClone.Name, len(pubClone.Status.DisruptedPods), len(pubClone.Status.UnavailablePods),
+			pubClone.Status.TotalReplicas, pubClone.Status.DesiredAvailable, pubClone.Status.CurrentAvailable, pubClone.Status.UnavailableAllowed)
+		if err = client.Status().Update(context.TODO(), pubClone); err == nil {
+			return nil
+		}
+		// if conflict, then retry
+		refresh = true
+		return err
+	})
+	if err != nil && err != wait.ErrWaitTimeout {
+		klog.Errorf("pod(%s.%s) operation(%s) for pub(%s.%s) failed: %s", pod.Namespace, pod.Name, operation, pub.Namespace, pub.Name, err.Error())
+		return false, err.Error(), nil
+	} else if err == wait.ErrWaitTimeout {
+		err = errors.NewTimeoutError(fmt.Sprintf("couldn't update PodUnavailableBudget %s due to conflicts", pub.Name), 10)
+		klog.Errorf("pod(%s.%s) operation(%s) failed: %s", pod.Namespace, pod.Name, operation, err.Error())
+		return false, err.Error(), nil
+	}
+
+	klog.V(3).Infof("admit pod(%s.%s) operation(%s) for pub(%s.%s)", pod.Namespace, pod.Name, operation, pub.Namespace, pub.Name)
+	return true, "", nil
+}
+
+func checkAndDecrement(podName string, pub *policyv1alpha1.PodUnavailableBudget, operation Operation) error {
+	if pub.Status.UnavailableAllowed < 0 {
```

Comment:
Question with regard to UnavailableAllowed:

1) How is the value updated? or initialized in the first place? 
2) Why do we need to persist the value in the status instead of calculating based on MaxUnavailable and MinAvailable in the Spec online?

Reply:
The amount of code submitted this time is relatively large, so the core code is divided into the following sections for submission.
1. Pub control: webhook and controller some public methods, as well as interfaces used to decouple the group and community implementation
2. webhook: used to intercept all pod requests and determine whether to allow the request according to UnavailableAllowed
3. controller: Calculate UnavailableAllowed based on the real situation of online pods (whether they are ready), etc. UnavailableAllowed is a complex calculation process and is not suitable to be calculated in webhook.
PodUnavailableBudgetValidatePod method should be a logical part of the webhook, but because it is intended to be called directly by in-place upgrade logic such as cloneset to achieve in-place upgrade protection, so it is placed in pub control